File: 599_MinimumIndexSumofTwoLists/599_MinimumIndexSumofTwoLists.js
/**
 * @param {string[]} list1
 * @param {string[]} list2
 * @return {string[]}
 */
var findRestaurant = function (list1, list2) {
  let cache = {};
  for (let i = 0; i < list1.length; i++) {
    cache[list1[i]] = i;
  }
  let min = Infinity;
  let ans = [];
  for (let i = 0; i < list2.length; i++) {
    if (list2[i] in cache) {
      cache[list2[i]] += i;
      if (min > cache[list2[i]]) {
        min = cache[list2[i]];
        ans = [list2[i]];
      } else if (cache[list2[i]] == min) {
        ans.push(list2[i]);
      }
    }
  }
  return ans;
};

File: 599_MinimumIndexSumofTwoLists/599_MinimumIndexSumofTwoLists.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        cache = defaultdict(int)
        for i, restaurant in enumerate(list1):
            cache[restaurant] = i
        
        lowest = float('inf')
        ans = []
    
        for i, restaurant in enumerate(list2):
            if restaurant in cache:
                total_index = cache[restaurant] + i
                if total_index < lowest:
                    lowest = total_index
                    ans = [restaurant]
                elif total_index == lowest:
                    ans.append(restaurant)
    
        return ans

        